### ISSUE TITLE
Find a way for a pre-commit hook to run faster (research)

### ISSUE THREAD

**Issue by lgajowy** [ScalaConsultants/mesmer#299]
Currently the [pre-commit](https://pre-commit.com/) hook is quite slow due to running sbt formatter and scalafix. It would be super useful to make it faster. 

Options to consider: 
 - run sbt natively somehow
 - use nailgun to reduce the jvm overhead: https://github.com/facebook/nailgun
 - ???

**Comment by lgajowy** [comment]
https://get-coursier.io/ might be something useful. It lets you install scalafmt and scalafix as commands that do not require running whole sbt with it. It's way faster this way. 
We could replace sbt fmt with scalafix; scalafmt commands in the pre-commit hook command. The only requirement is that devs have to install it using cursier (not a big deal).

**Comment by lgajowy** [comment]
@worekleszczy this is what we were talking about btw. ;)

**Comment by worekleszczy** [comment]
@lgajowy I research running scalafix from `coursier` and it's not going to solve our problems - it is possible to run it even with third party rules, but for semantics rules we need `semanticdb` run over all files and keeping those 2 tools in sync would be counter productive. Another thing is that we'd require to set up whole classpath for it with all dependencies and and the only place we have this information is in `sbt`.

**Comment by lgajowy** [comment]
That's a shame. 

Maybe we should try this to speed things up https://scalacenter.github.io/bloop/docs/what-is-bloop? Not only scalafix/scalafmt but build time in general. In my head it's a similar gain as it would be with nailgun. Not sure which tool is better but we should research :)

**Comment by worekleszczy** [comment]
Nailgun seems promising, but for me this is a `nice to have` - I wouldn't spend to much time investigating this. Using `coursier` seemed like a low hanging fruit, but after learning a little bit more about it I see that this was pretty naive - to perform many of the `scalafix` rules we need to compile the whole codebase and this is the most time consuming part.

**Comment by lgajowy** [comment]
I tried using bloop but not much luck. However, I have an interesting finding: when I run `bloop test otelExtension` the tests timeout constantly. No idea why.

I also tried this: https://www.scala-sbt.org/1.x/docs/sbt-1.4-Release-Notes.html#Native+thin+client and it looks a lot more promising. Locally, it's enough to set an env variable: 

```export SBT_NATIVE_CLIENT=true```

it will pass all the compile requests to a build server process and hence will speedup the pre-commit (and testing, and compilation in general as an added value). Note it's marked as "experimental" but IMHO not a big deal - we can always turn off the flag.
